### ISSUE TITLE
Linking libzstd statically

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/ocaml#12562]
I’ve only now realized that the new libzstd optional dependency is linked dynamically. https://github.com/ocaml/ocaml/issues/12431 shows that mess ups by the system package manager is not uncommon but beyond that I’m rather afraid of seeing a flock of users pretty soon complaining that their new ocaml install do not work because homebrew removed libzstd or whatever.

As a way of counteracting that, would it not be possible to link libzstd statically by default instead so that such breakage does not happen?

**Comment by glondu** [comment]
Probably. This may be done by putting `-lzstd` between `-Bstatic` and `-Bdynamic`, or by directly refering to the `.a` file instead of `-lzstd`. But please make it optional.

**Comment by xavierleroy** [comment]
No, we're not going to statically link libzstd in every executable produced by ocamlopt, if only because the code of libzstd is TWICE AS BIG as the whole OCaml runtime system: 832k vs 376k on x86-64.  Don't ask me why so much code...

The Brew package for libzstd was fixed and now points to the permanent location /opt/homebrew/opt/zstd .  For Linux distributions, I anticipate no more difficulties with libzstd.so than with libc.so.  Am I missing a possible problem?

**Comment by glondu** [comment]
Every executable produced by ocamlopt needs libzstd?!? :thinking:

**Comment by xavierleroy** [comment]
> Every executable produced by ocamlopt needs libzstd?!?

I'm afraid so.  ZSTD compression is available through the standard library module Marshal, so libzstd functions are referenced from the runtime system.  Which is not a problem if it's linked dynamically, but an (unexpected) concern if it is linked statically.

**Comment by kit-ty-kate** [comment]
This is a huge change for distribution packagers that now have to track down every single packages that use OCaml and add this dependency.

> I anticipate no more difficulties with libzstd.so than with libc.so. Am I missing a possible problem?

I disagree. On a single machine libc.so is guaranteed to always be there, whereas libzstd can be uninstalled at will. This is a big problem for opam that does not manage the system package manager (only install things if needed), so if your system package manager or you decides to uninstall (or break) libzstd for whatever reason all your OCaml programs suddenly do not work anymore. For simple users without deep knowledge of the compiler this is going to be really confusing and problematic.

**Comment by gasche** [comment]
The zstd README documents various build-time options ([README: modular build](https://github.com/facebook/zstd/blob/dev/lib/README.md#modular-build)), many meant to reduce the size of the library artifact, in particular:
- ZSTD_LIB_MINIFY which asks to systematically reduce code size at a potential performance cost
- options to specialize the build for only certain compression strategies

It we want to consider static linking as an option, it would be very useful if someone played with this a bit. Can we choose a set of option that results in a much smaller library, without making compression too slow in practice?
(I'm not sure how we would actually go about deploying a custom build of zstd for use in the compiler. This could be an opam package that the compiler depends on, but then build time becomes relatively critical.)

**Comment by jonahbeckford** [comment]
(Please take my comments with a grain of salt since I'm not immediately impacted by this OCaml 5 only issue)

> Am I missing a possible problem?

TLDR: On Windows the libzstd change will impact ordinary `ocamlopt` generated executables. I'm providing some context, but not any recommendations.

On Windows any DLLs need to be in the PATH or co-located with the ocamlopt generated executables.

#### Yesterday

For MSVC created executables only `vcruntime140.dll` is needed. Almost every Windows Administrator installer auto-installs it into `C:\Windows\System32\vcruntime140.dll` so it automatically is in the PATH. Most Windows users (and users of `ocamlopt` generated binaries) are unaware of that dependency except occasionally a brand-new PC may encounter problems. See my troubleshooting page https://github.com/diskuv/dkml-workflows#distributing-your-executable for more details of that edge case.

For GCC created executables there are a couple DLLs that need to be present: `libgcc_s_seh-1.dll` and usually `libwinpthread-1.dll`. These are not present on a Windows machine, so the only practical choice is to co-locate the DLLs alongside the executable. For example, if you look at the latest version of Unison at https://github.com/bcpierce00/unison/releases/tag/v2.53.3 (unison-2.53.3-windows-x86_64.zip) you will see:
```text
    Directory: C:\Users\beckf\Downloads\unison-2.53.3-windows-x86_64\bin

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         9/16/2023   6:16 AM          33792 iconv.dll
-a----         9/16/2023   6:16 AM         128000 libatk-1.0-0.dll
-a----         9/16/2023   6:16 AM          71168 libbz2-1.dll
...
-a----         9/16/2023   6:16 AM          31744 libffi-6.dll
...
-a----         9/16/2023   6:16 AM          80384 libgcc_s_seh-1.dll
...
-a----         9/16/2023   6:16 AM          55827 libwinpthread-1.dll
-a----         9/16/2023   6:16 AM        1557504 unison-fsmonitor.exe
-a----         9/16/2023   6:16 AM        6574080 unison-gui.exe
-a----         9/16/2023   6:16 AM        6574080 unison-text+gui.exe
-a----         9/16/2023   6:16 AM        3473920 unison.exe
-a----         9/16/2023   6:16 AM         117267 zlib1.dll
```

#### Today with the zstd change

Unison (GCC), my own [diskuvbox (MSVC)](https://github.com/diskuv/diskuvbox/releases), and every other `ocamlopt` generated executable will need to place `libzstd.dll` in the same directory as the executable.

The problem is worse for MSVC because typically just executing the `.exe` is all that is required for 99.9% of Windows users (because `vcruntime140.dll` is almost always present from a previous installer). GCC users have to bundle a bunch of DLLs already ... one more should not be much of a problem as long as it is communicated.

**Comment by gasche** [comment]
Note that any distribution of OCaml can choose to disable Marshal's compression support at configure-time. There is no dependency on zstd then, and programs that try to use compression during marshalling will still work (the flag is silently ignored in this case). The only potential issue is if they receive a compressed marshalled payload coming from another system, which they would be unable to uncompress. (They would also obviously not benefit from the compression of compiler build artifacts, so for example their `opam` installation would be noticeably larger.)

**Comment by jonahbeckford** [comment]
> Note that any distribution of OCaml can choose to disable Marshal's compression support at configure-time.

Just to be clear, the implication is the recommendation that system packagers and anyone distributing binaries to the public use a `./configure --disable-zstd` (whatever the flag is) to produce their `ocamlopt`? At least for the people who distribute individual binaries (ex. change the `setup-ocaml` GitHub Actions to disable that flag) that could be a workable solution.

**Comment by dbuenzli** [comment]
I don't think the problem will be with distributions, they know how to handle and record their dependencies.

I think the opam experience though will be a problem because you install new compilers which will pickup either the presence or absence of libzstd in the environment without that dependency to be formally registered anywhere.

So I think this comment by @kit-ty-kate :

> I’m rather afraid of seeing a flock of users pretty soon complaining that their new ocaml install do not work because homebrew removed libzstd or whatever.

is real. One way of avoiding this would be to:

1. (optional) Ask system packagers of opam to make it dependent on `zstd`.
2. Make the `ocaml` opam package depend on the `conf-zstd` opam package.

That would minimize the possibility of these unregistred dependencies to be broken. (But would force `zstd` on all opam users).

**Comment by gasche** [comment]
@jonahbeckford I was not trying to recommend, but to point out to concerned distribution-makers that this option exists. (In the sense of "OCaml distribution", not "Linux distribution") There are many different sorts of distributions (opam, Linux distros, Homebrew, your dkml work, Docker images providers, etc.), and they certainly want different tradeoffs among linking zstd dynamically, statically, or disabling it at configure time to avoid the hassle.

I agree with @dbuenzli:
- linux distribution "system packages" can trust that they have a full view of dependencies (and they tend to prefer dynamic linking), it makes sense for them to support a dynamically-linked zstd
- opam is weird, because the experience depends on whether you use the system-provided `opam` package (which can easily enforce a `libzstd` dependency) or you download an opam binary from the web (cannot). The opam-repository has to deal with both users, so some in-repository logic to detect the status makes sense.
- if dynamic linking is too painful or fragile in practice for Windows distributions, or Homebrew, etc., those distributions could link statically or disable zstd support in the OCaml runtime

I think it would make sense to add optional support for static linking of libzstd, not enabled by default, to make this choice easy for distributions that prefer it. In the meantime, they should consider disabling the option.

(My earlier idea of tuning the compilation options for libzstd static linking sounds good in theory but it means we have to build zstd ourselves, which is probably a terrible idea in practice, especially if we only try this on those platforms which provide bad development environments by default.)

**Comment by alainfrisch** [comment]
> Unison (GCC), my own [diskuvbox (MSVC)](https://github.com/diskuv/diskuvbox/releases), and every other ocamlopt generated executable will need to place libzstd.dll in the same directory as the executable.

Just as a side note, it seem you'll also need to include the libzstd's BSD LICENSE file in the distrib (it's dual licensed with both BSD and GPLv).

Cannot we arrange so that ocaml/ocamlopt benefit from libzstd when the build machine has the library available, but without creating a strong dependency on it for executables built with ocamlopt, perhaps using lazy loading for the .so/.dll?

**Comment by kit-ty-kate** [comment]
> ZSTD_LIB_MINIFY which asks to systematically reduce code size at a potential performance cost

I tried it locally and it shrunk the size of `libzstd.a` from 807K to 352K. The other options did not reduce it by much more.

Is that too big to add to every OCaml binaries? If so, would it sound reasonable to split the `Marshal` module in two (one builtin, and the other with compression support) similar to the `bigarray` library back in 4.07

**Comment by dra27** [comment]
> Is that too big to add to every OCaml binaries? If so, would it sound reasonable to split the `Marshal` module in two (one builtin, and the other with compression support) similar to the `bigarray` library back in 4.07

It's not quite the same problem - with the separate bigarray library, that was providing two different _OCaml_ modules to access the same underlying runtime features. Here, we want to have the runtime lack ZSTD by default, but be able to add ZSTD support on demand.

However, I think it should be relatively easy to do something in intern and extern akin to the runtime hooks used by both systhreads and unix and move all the ZSTD-related intern/extern code to a separate library. Unless you link that new library, `Marshal.compression_supported ()` will return `false`. The fiddly part is that this needs to be plumbed in, a la `otherlibs/systhreads/thread.ml` is: https://github.com/ocaml/ocaml/blob/ad868faff091eae242ee5625f9f8764a3cce5869/otherlibs/systhreads/thread.ml#L76-L77
 which either means coming up with an OCaml module name (which will never be used), or slightly extending cm(x)a to permit expunge-style linking of modules where only their toplevel statements are linked, but the name itself doesn't become available.

I think @alainfrisch's idea runs a risk of being unexpectedly brittle. If the user knowinglycompiled a program _with_ zstd support, I think it's better that there's then definitely an error if it's run somewhere which doesn't have that zstd support, rather than quietly disabling it, especially given that the `Compression` flag is silently ignored if compression isn't available?

**Comment by xavierleroy** [comment]
FWIW, I have an implementation of the dynamic loading approach suggested by @alainfrisch here: https://github.com/xavierleroy/ocaml/tree/dlopen-zstd

> If the user knowinglycompiled a program with zstd support, I think it's better that there's then definitely an error if it's run somewhere which doesn't have that zstd support, rather than quietly disabling it, especially given that the Compression flag is silently ignored if compression isn't available?

If OCaml was configured with ZSTD support and libzstd.so cannot be loaded, we could raise an exception with a clear error message rather than switching to uncompressed output.

> However, I think it should be relatively easy to do something in intern and extern akin to the runtime hooks used by both systhreads and unix and move all the ZSTD-related intern/extern code to a separate library.

I thought about this too, but had the impression that it's more code and a more complicated build.  Unlike systhreads, compressed marshaling must be available to the ocamlc and ocamlopt compilers.

**Comment by dra27** [comment]
I've experimented, too - all with no definite conclusion in either direction. A few thoughts/musings:
- Moving the zstd-specific parts of `extern.c` and `intern.c` to a separate library was less invasive than I'd expected - it's only `struct output_block` which I needed to be move to `intext.h`; the other handling can be done by having a return status on the decompression and compression functions.
- Making that available to the `.opt` versions of the compilers is trivial (it's just an extra .cmxa, along with the compiler-libs)
- Making that available during the bytecode parts of the build is obviously less-trivial, but isn't _too_ hairy as long as we start with the reasonable assumption that the binary interpreters (i.e. `ocamlrun`, `ocamlrund`, `ocamlruni`) are compiled with zstd support.
- The final result of my experiment was an additional package (for now called `stdlib.marshal-zstd`) which user-code should add to `ocamlfind` invocations / Dune `libraries` stanzas if zstd support is required in programs, but the compilers have it unconditionally. The `META` file for `compiler-libs` references it, so any users of ocamlcommon (who may therefore be loading our binary artefacts) would seamlessly get the decompression support.

I've got three worries with the `dlopen` approach, but none of them is insurmountable:
- I fear there's a devil in the detail of exactly which DLL gets loaded at runtime. That's definitely an issue on Windows, and I imagine that macOS is capable of making this hard... with `-lzstd`, we leave a lot to the linker to work out exactly what that means.
- I'm clearly not a lawyer, but the suggestion reading around is that switching to use `dlopen` as a means of bypassing license restrictions doesn't work. If I'm reading it correctly, binaries compiled with OCaml 5.1+ when zstd support is enabled _must_ distribute the [zstd copyright notice](https://github.com/facebook/zstd/blob/dev/LICENSE), unlike with our LPGL+linking-exception licence which only requires the copyright _if_ the program already has a copyright notice. This can of course be trivially circumvented by using switches with zstd disabled, but it does create a small incentive to turn the compression off completely.
- In switching to `dlopen`, we're making compilation in static environments harder (`opam`'s binary distributions are compiled statically and consumed that way by quite a lot of OCaml users; opam also stores a large `Marshal`'d cache under `~/.opam`)

**Comment by kit-ty-kate** [comment]
> In switching to `dlopen`, we're making compilation in static environments harder

Some platforms do not even have a static version of zstd available (e.g. https://archlinux.org/packages/core/x86_64/zstd/) so the expectation that worked before that you can just take any OCaml installation and create a static binary with it doesn't hold true anymore with 5.1.0. Now users who wants to do that have to either vendor OCaml and compile it with `--without-zstd` or vendor zstd and compile it themselves just in case the current OCaml version happens to be 5.1.0.

**Comment by xavierleroy** [comment]
> the expectation that worked before that you can just take any OCaml installation and create a static binary with it 

I was under the impression that Glibc-based systems are not really able to produce static binaries, and have had problems with that for the last 20 years, and that Musl must be used to get static linking, in which case #12564 is relevant.

At any rate, besides obsessing over static linking of ZSTD, could someone summarize the license issues?  That's something that went COMPLETELY UNNOTICED during the review of #12006.

**Comment by xavierleroy** [comment]
> I've experimented, too - all with no definite conclusion in either direction. 

We should compare notes about the "extra library" approach.  If it doesn't complicate our build system too much, I'm open to it.  But then I would advocate for a change of API, with a separate `Compress` module or `ZSTD` module providing compressed marshaling and a few other compression/decompression functions.  The approach in 5.1 was completely based on the idea that compression is handled as transparently as possible, even when not available.  If an extra library must be linked to get compression, it's a lot less transparent, and `Marshal` is probably the wrong place to expose it.

Concerning the license issues, I'm still very interested  in a summary of the jurisprudence concerning dynamic linking and advertising clauses in licenses like the 3-clause BSD.

**Comment by dbuenzli** [comment]
> Concerning the license issues, I'm still very interested in a summary of the jurisprudence concerning dynamic linking and advertising clauses in licenses like the 3-clause BSD.

I have no idea about the jurisprudence but note that it's the [4-clause BSD](https://spdx.org/licenses/BSD-4-Clause.html) that has the controversial *advertising* (as in ad) clause. 

The [3-clause](https://spdx.org/licenses/BSD-3-Clause.html) simply says that you need to mention the license in the docs of the software – similar to the requirements of 6. in the LGPL that OCaml nicely allows to eschew under some conditions.

It's a bit annoying and definitively a dent in the nice terms that OCaml is provided under. OTOH I suspect that nowadays any software product likely eventually depends on some free software whose licence needs to be mentioned in the docs/about dialog.

**Comment by xavierleroy** [comment]
All right, the BSD license used by ZSTD has 3 clauses.  What are the implications for dynamic loading of this library?  I really need to be explained like I'm 6 years old.

**Comment by dbuenzli** [comment]
> What are the implications for dynamic loading of this library?

I have no answer to this question but then I'm not sure it's possible to answer it in the abstract. What is your use case ? Who provides the zstd dll ? 

Suppose for example I make an app for an app store using OCaml. Whether I link dynamically or statically against zstd is a bit of a moot point: in either case my app bundle is going to redistribute zstd's code in binary format: whether as a dll in my app bundle or statically linked in the executable of the bundle. So I will have to add the zstd licence in the supporting material of the app (about dialog, docs) to satisfy the second clause of the BSD3.

If your app dynamically links against say a system provided zstd dll. Then I have no idea but I suspect you don't have to mention it (you are not the one "redistributing" the code but IANAL).

**Comment by kit-ty-kate** [comment]
> Suppose for example I make an app for an app store using OCaml. Whether I link dynamically or statically against zstd is a bit of a moot point: in either case my app bundle is going to redistribute zstd's code in binary format: whether as a dll in my app bundle or statically linked in the executable of the bundle. So I will have to add the zstd licence in the supporting material of the app (about dialog, docs) to satisfy the second clause of the BSD3.

The OCaml compiler could always include the required BSD3 license in every OCaml binary as a static global string with ` __attribute__((used))` (unstrippable).

I’m not a lawyer but it looks to me like this should fulfil clause number 2 of the BSD3 license just fine if we went the static linking route too.

**Comment by dbuenzli** [comment]
Whether legally right or wrong I personally don't find what you suggest to be in the spirit of the license which asks to provide credits for the software in end-user docs. I would find it disrespectful to the authors of the software.

**Comment by aryx** [comment]
Can someone create urgently a configure --disable-std and an opam switch 5.1.0+nozstd? @kit-ty-kate ?

The current situation is pretty annoying for people trying to distribute static binaries of their OCaml program (for example we do that for semgrep but can't anymore update to 5.1.0 because of zstd: https://github.com/returntocorp/semgrep/pull/8922 )

**Comment by gasche** [comment]
@Octachron has proposed to create an `ocaml-option-no-compression` package to make it easy for users to disable zstd support in their OCaml install. I think that would solve your problem (and others).

@Octachron appears to be rather busy with this and other last-minute 5.1 issues. Anyone else should feel free to contribute this to opam-repository. It should be a fairly simple change given that the compiler already has configure-time support (we *already* have a `--without-zstd` option).

**Comment by dra27** [comment]
We're on it. As a temporary workaround, if the static binaries are being built in Docker or another CI system, is it not possible just to ensure that libzstd-dev (or equivalent) is not installed? I'm not clear why why the dependency on conf-zstd is being added in the linked semgrep PR.

**Comment by aryx** [comment]
Don't look at the semgrep PR; we're experimenting with stuff. The main issue we have is that on macos we can't easily statically link; we have to do some ugly tricks like erasing the .dylib to force the linker to use the .a, but then while it worked for the pcre and gmp library, we can't make it work with zstd because the moment we remove libzstd.1.dylib, we can't even use ocamlc anymore ...

**Comment by aryx** [comment]
So the simplest path for us is probably a special opam switch that disable zstd.

**Comment by avsm** [comment]
xrefing https://github.com/ocaml/opam-repository/pull/24608 here.

**Comment by dra27** [comment]
Ah, that makes complete sense to me now, @aryx, thanks!

**Comment by avsm** [comment]
@aryx the opam-repo PR is now merged, so you should be able to:

```
$ opam switch create 5.1.0-nozstd --packages ocaml-variants.5.1.0+options,ocaml-option-no-compression 
```

or a variant with musl.

**Comment by xavierleroy** [comment]
See proposed remediation at #12705 .

**Comment by UnixJunkie** [comment]
We could also have an opam package that installs libzstd under opam's control (not relying on whatever is installed on the system).
For ocaml versions where Marshal supports compression via libzstd, we make this package a required dependency.

**Comment by kit-ty-kate** [comment]
related issue: https://github.com/ocaml/ocaml/issues/12835

**Comment by kit-ty-kate** [comment]
thinking about this issue again, I’m wondering how feasable it is to always ship libzstd with OCaml. This way the ocaml distribution controls its runtime dependencies and this should eliminates most of the downsides.
I can’t remember if it was discussed before, if so please ignore me it’s probably a bit too late in the night.

**Comment by UnixJunkie** [comment]
I was also suggesting something a bit similar previously.

**Comment by UnixJunkie** [comment]
If this marshal compression thing comes back, would it be possible to have the compression level as an optional parameter (w/ default value 3) so that users can tweak it?
E.g. if you know you are compressing a very large file, using 'zstd -1' makes sense.
